README.md

```diff
@@ -21,14 +21,45 @@ Once you have implemented the `fm` program (and tested it to the best of your ab
+We store all the necessary information in a class called fm index, this includes the Suffix array, the string (as a rotating string), a mapping between alphabet & index, and the O and C tables. 
+The pre processing calculates these and dumps them to a binary file using pickle.
+
+The data necessary data structures should follow these complexities:
+* Suffix array - O(n log n) no, we don't have a linear implementation yet...
+* Rotating string - O(n), unfortunately we make a copy of the input string, but hey this was easier
+* O table - O(n) given that we can assume a constant sized alphabet, as the table is sized n * $\sigma$ 
+* C table - O(n) we do one pass over the alphabet, one pass over the string for counts. we then celebrate passover.
+* Alphabet mapping - O(n) one iteration on the alphabet...
+
+That linear time SA construction is looking very appealing right about now...
+
+
+
+
 ## Insights you may have had while implementing the algorithm
+We played around abit with RLE and the other tricks related to the bwt, so if we wanted to reduce the size of our pre-processed data, we could drop the string, and replace it with the bwt l-column compressed with rle, which would save us a bit. but that would then cost in loading time. On the same note we could cut out the suffix array, as that can be reconstructed in reasonable time (we atleast convinced ourselves that it should be possible in O(n) using the O table). It looks like there is ample opportunity for playing around with compression.
 
 ## Problems encountered if any
+We didn't encounter any soulcrushing problems, only minor soulcrushing hiccups. 
+Jokes aside, this proved to be one of the easier implementations we've done, so as far as we're aware we didnt encounter any major problems
 
 ## Validation
 
 *How did you validate that the preprocessing and the search algorithm works?*
+Once again, for searching we can compare results with our earlier implementations. 
+Generally while testing if the algorithm worked we ensured that it would handle empty strings as both x and pattern, aswell as single letter, and repetetive strings. So far none of these break.
+
+The pre-processing was tested by dumping an object to file and reloading, then comparing with a newly constructed to see if anything went missng, and by running pre-processing + search in the same vs in seperate scripts.
 
 ## Running time
 
 *List experiments and results that show that both the preprocessing algorithm and the search algorithm works in the expected running time. Add figures by embedding them here, as you learned how to do in project 1.*
+
+Running time experiments for preprocessing (in our case creating the fm_index object) show something that looks like a $linear * log$ time curve, which is as expected. The sudden outliers in the end is probably just me checking Facebook when I got tired of waiting. Here, we go through 100 strings from 1000 to 100000 in length.
```

Comment:
How can I tell that this graph shows $O(n\log n)$ and not, for example, $O(n\sqrt{n})$ or $O(n\log\log n)$?

If it is not a line, divide the measured time by the expectation to convince me.